src/dumpfile_parse.rs

```diff
@@ -0,0 +1,867 @@
+//! # Parsing and generating composefs dump file entry
+//!
+//! The composefs project defines a "dump file" which is a textual
+//! serializion of the metadata file.  This module supports parsing
+//! and generating dump file entries.
+use std::borrow::Cow;
+use std::ffi::OsStr;
+use std::ffi::OsString;
+use std::fmt::Display;
+use std::fmt::Write as WriteFmt;
+use std::fs::File;
+use std::io::BufRead;
+use std::io::Write;
+use std::os::unix::ffi::{OsStrExt, OsStringExt};
+use std::path::{Path, PathBuf};
+use std::process::Command;
+use std::str::FromStr;
+use std::usize;
+
+use anyhow::Context;
+use anyhow::{anyhow, Result};
+use rustix::fs::FileType;
+
+/// https://github.com/torvalds/linux/blob/47ac09b91befbb6a235ab620c32af719f8208399/include/uapi/linux/limits.h#L13
+const PATH_MAX: u32 = 4096;
+/// Maximum size accepted for inline content.
+const MAX_INLINE_CONTENT: u16 = 5000;
+/// https://github.com/torvalds/linux/blob/47ac09b91befbb6a235ab620c32af719f8208399/include/uapi/linux/limits.h#L15
+/// This isn't exposed in libc/rustix, and in any case we should be conservative...if this ever
+/// gets bumped it'd be a hazard.
+const XATTR_NAME_MAX: usize = 255;
+// See above
+const XATTR_LIST_MAX: usize = u16::MAX as usize;
+// See above
+const XATTR_SIZE_MAX: usize = u16::MAX as usize;
+
+#[derive(Debug, PartialEq, Eq)]
+/// An extended attribute entry
+pub struct Xattr<'k> {
+    /// key
+    pub key: Cow<'k, OsStr>,
+    /// value
+    pub value: Cow<'k, [u8]>,
+}
+/// A full set of extended attributes
+pub type Xattrs<'k> = Vec<Xattr<'k>>;
+
+/// Modification time
+#[derive(Debug, PartialEq, Eq)]
+pub struct Mtime {
+    /// Seconds
+    pub sec: u64,
+    /// Nanoseconds
+    pub nsec: u64,
+}
+
+/// A composefs dumpfile entry
+#[derive(Debug, PartialEq, Eq)]
+pub struct Entry<'p> {
+    /// The filename
+    pub path: Cow<'p, Path>,
+    /// uid
+    pub uid: u32,
+    /// gid
+    pub gid: u32,
+    /// mode (includes file type)
+    pub mode: u32,
+    /// Modification time
+    pub mtime: Mtime,
+    /// The specific file/directory data
+    pub item: Item<'p>,
+    /// Extended attributes
+    pub xattrs: Xattrs<'p>,
+}
+
+#[derive(Debug, PartialEq, Eq)]
+/// A serializable composefs entry.
+///
+/// The `Display` implementation for this type is defined to serialize
+/// into a format consumable by `mkcomposefs --from-file`.
+pub enum Item<'p> {
+    /// A regular file
+    Regular {
```

Comment:
This is an insufficient representation for regular files: they also need to be able to carry a payload in order to reference the backing file.  See https://man.archlinux.org/man/composefs-dump.5.en#PAYLOAD